rust/src/http2/parser.rs

```diff
@@ -344,39 +357,38 @@ fn http2_parse_headers_block_indexed<'a>(
     input: &'a [u8], dyn_headers: &HTTP2DynTable,
 ) -> IResult<&'a [u8], HTTP2FrameHeaderBlock> {
     fn parser(input: &[u8]) -> IResult<&[u8], (u8, u8)> {
-        bits!(
-            input,
-            complete!(tuple!(
-                verify!(take_bits!(1u8), |&x| x == 1),
-                take_bits!(7u8)
-            ))
-        )
+        bits::<_, _, Error<(&[u8], usize)>, _, _>(complete(tuple((
+            verify(take_bits(1u8), |&x| x == 1),
+            take_bits(7u8),
+        ))))(input)
     }
     let (i2, indexed) = parser(input)?;
     let (i3, indexreal) = http2_parse_var_uint(i2, indexed.1 as u64, 0x7F)?;
     if indexreal == 0 && indexed.1 == 0x7F {
-        return Err(Err::Error((i3, ErrorKind::LengthValue)));
+        return Err(Err::Error(make_error(i3, ErrorKind::LengthValue)));
     }
     match http2_frame_header_static(indexreal, dyn_headers) {
         Some(h) => Ok((i3, h)),
-        _ => Err(Err::Error((i3, ErrorKind::MapOpt))),
+        _ => Err(Err::Error(make_error(i3, ErrorKind::MapOpt))),
     }
 }
 
 fn http2_parse_headers_block_string(input: &[u8]) -> IResult<&[u8], Vec<u8>> {
     fn parser(input: &[u8]) -> IResult<&[u8], (u8, u8)> {
-        bits!(input, tuple!(take_bits!(1u8), take_bits!(7u8)))
+        let (i, b) = be_u8(input)?;
+        Ok((i, (b >> 7, b & 0x7f)))
     }
     let (i1, huffslen) = parser(input)?;
     let (i2, stringlen) = http2_parse_var_uint(i1, huffslen.1 as u64, 0x7F)?;
     if stringlen == 0 && huffslen.1 == 0x7F {
-        return Err(Err::Error((i2, ErrorKind::LengthValue)));
+        return Err(Err::Error(make_error(i2, ErrorKind::LengthValue)));
     }
-    let (i3, data) = take!(i2, stringlen as usize)?;
+    let (i3, data) = take(stringlen as usize)(i2)?;
     if huffslen.0 == 0 {
         return Ok((i3, data.to_vec()));
     } else {
-        let (_, val) = bits!(data, many0!(huffman::http2_decode_huffman))?;
+        let (_, val) =
+            bits::<_, _, Error<(&[u8], usize)>, _, _>(many0(huffman::http2_decode_huffman))(data)?;
```

Comment:
this annotation is not so nice.
How did you find it ?

Reply:
Agreed, it is neither nice nor readable.

The rust compiler is lost in generic types in the `bits` combinator, and fails to infer the error type correctly. Here, the solution is to explicit the error type while letting the compiler infer other generic parameters (`_`).
This problem did not exist with macro (but they had other problems), and seem to happen only with the `bits` combinator. The only way I found to avoid this is to create a function containing the subparser, so the error type is given in the function declaration and the compiler infers the rest - but this is not great, because it prevents inlining the parser ...

The annotation is similar to the one used in nom's documentation: https://docs.rs/nom/7.1.0/nom/bits/fn.bits.html

Reply:
Maybe this deserves a comment...

Reply:
What prevents the inlining? Its used enough that a helper function could clean it up a little.

Reply:
Nothing prevents inlining (and we can use the `#[inline]` attribute to force it). The size of the bits seems to vary, so it is possible to write a function that would take some parameters as arguments and return a combinator.
If that's ok for you, I will push the patch to the end of this PR.